Automattic/newspack-plugin
# feat: add GA4 donation events

### All Submissions:






### Changes proposed in this Pull Request:



Connects the Donation events to GA4.

### How to test the changes in this Pull Request:

Get the GA4 credentials. You will need your measurement ID and an API secret.

* api_secret - Required. An API SECRET generated in the Google Analytics UI. To create a new secret, navigate to:
    Admin > Data Streams > choose your stream > Measurement Protocol > Create
* measurement_id - Required. The measurement ID associated with a stream. Found in the Google Analytics UI under:
    Admin > Data Streams > choose your stream > Measurement ID

Store this info in the database:
```
wp option set ga4_measurement_id "G-XXXXXXXXXX"
wp option set ga4_api_secret YYYYYYYYYYYYYYYYYYY
```
Now, let's test:

1. Make sure RAS is active and that you have a registered reader
2. Make sure to add this constant to wp-config
```
define( 'NEWSPACK_LOG_LEVEL', 2 );
define( 'NEWSPACK_EXPERIMENTAL_GA4_EVENTS', true );
```
4. Make a donation to the site via the donation block.
5. Watch the `donation_new` event being fired in the logs
```
[NEWSPACK-DATA-EVENTS][Newspack\Data_Events::dispatch]: Dispatching action "donation_new" via Action Scheduler.
```
6. Watch the handler being triggered and sending the `reader_registered` GA event:
```
[NEWSPACK-DATA-EVENTS][Newspack\Data_Events::handle]: Executing action handlers for "donation_new".
[NEWSPACK-DATA-EVENTS-GA4][Newspack\Data_Events\Connectors\GA4::log]: Event sent - donation_new
```
7. Visit the Google Analytics Realtime dashboard
8. See the `donation_new` events coming in
9. Inspect them and see check its attributes to see if the values are correct.

![Captura de tela de 2023-03-07 08-37-09](https://user-images.githubusercontent.com/971483/223423517-d133d890-88cd-4681-9380-c2a0e7cf4cdf.png)

10. In the admin panel, go to Woocommerce > Subscriptions
11. Cancel the subscription
12. See that a `donation_subscription_cancelled` event will be fired
13. Check all its metadata
14. Repeat steps 5 to 13  using both Stripe and Newspack as Reader Revenue platforms

NOTE: `referer` and `popup_id` attributes are being added by https://github.com/Automattic/newspack-plugin/pull/2310

### Other information:
